src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs

```diff
@@ -209,38 +177,107 @@ private static string ExtractDatabaseNameForLogging(TcpConnectionOptions tcpConn
             }
         }
 
-        protected virtual void OnError(Exception e) { }
+        internal virtual async Task HandleBatchStatus<TState, TConnection>(TState state, BatchStatus status, Stopwatch sendingCurrentBatchStopwatch, DisposeOnce<SingleAttempt> markInUse,
+            SubscriptionBatchStatsScope batchScope) where TState : AbstractSubscriptionConnectionsState<TConnection, TIncludesCommand>
+            where TConnection : SubscriptionConnectionBase<TIncludesCommand>
+        {
+            switch (status)
+            {
+                case BatchStatus.EmptyBatch:
+                    await LogBatchStatusAndUpdateStatsAsync(sendingCurrentBatchStopwatch, $"Got '{nameof(BatchStatus.EmptyBatch)}' for subscription '{Options.SubscriptionName}'.");
+
+                    if (FoundAboutMoreDocs())
+                        state.NotifyHasMoreDocs();
 
-        protected List<(Document Document, Exception Exception)> CurrentBatch = new();
+                    AssertCloseWhenNoDocsLeft();
+
+                    if (await WaitForChangedDocsAsync(state))
+                        return;
+
+                    await CancelSubscriptionAndThrowAsync();
+
+                    break;
+
+                case BatchStatus.DocumentsSent:
+                    markInUse.Dispose();
```

Comment:
why we are disposing this here? the outer using is not enough?

Reply:
in previous code the `WaitForClientAck` happens after markInUse is disposed:
https://github.com/ravendb/ravendb/blob/3b0f69bd2a70dc5a831eb5a49467e77087b43217/src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs#L193